### PR TITLE
Accelerometer and Gyroscope support on mobile devices

### PR DESCRIPTION
**Pull request by mulshine**
`Accel` and `Gyro` classes implemented for WebChucK to be used on mobile devices. Both modeled after `HID`,`Accel` 
 and `Gyro` interface with browser/JS "devicemotion" and "deviceorientation" events, respectively, and allow you to listen for and receive `AccelMsg`'s or `GyroMsg`'s inline with WebChucK code, providing x, y, z data in each case. 

```
Gyro gy;
GyroMsg msg;

0 => int device;

// open gyro 
if( !gy.openGyro( device ) ) me.exit();
<<< "gyro '" + gy.name() + "' ready", "" >>>;

<<< "only on mobile" >>>;

// infinite event loop
while( true )
{
    // wait on gyro event
    gy => now;

    // get one or more messages
    while( gy.recv( msg ) )
    {
        // print gyro values
        <<< msg.getGyroX() + " " + msg.getGyroY() + " " + msg.getGyroZ() >>>;
    }
}
```

Big thanks to @terryzfeng for assistance in the architecture here!

### CODE REVIEW

File: src/gyro/Gyro.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
+import Chuck from "../Chuck";
+import { Gyro_ck, GyroMsg_ck } from "./gyroCk";
+
+
+/**
+ * Introducing Gyro (gyroerometer, on mobile) support for WebChucK. Gyro wraps
```
Comment (terryzfeng) [inline src/gyro/Gyro.ts:6]: nit: Do you mean gyrometer?

File: src/gyro/Gyro.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * const gyro = await Gyro.init(theChuck); // Initialize Gyro
+ * ```
+ * 
+ * The "deviceorientation" event gives motion of the device around the three axes (x, y, and z) represented as degrees from 0 to 360.
+ * More on the "deviceorientation" event can be found online {@link https://developer.mozilla.org/en-US/docs/Web/API/Window/deviceorientation_event | here }.
+ * 
+ * iOS devices require that the web developer ask permission from the user to access sensors after a button push. This looks like: 
+ * 
+ * ```ts
+ * import { Chuck, Gyro} from '../webchuck/src/wc-bundle.js';
```
Comment (terryzfeng) [inline src/gyro/Gyro.ts:25]: When mentioning importing for documentation, you can just say `import { Chuck, Gyro } from "webchuck"`. The location is different on every platform so it's easier to keep it general.

Nit: Can you make sure there's a opening and closing space between `{ Chuck, Gyro }`

File: src/gyro/Gyro.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    // Initialize members
+    this.theChuck = theChuck;
+
+    this.boundHandleOrientation = this.handleOrientation.bind(this);
+  }
+
+  /**
+   * Initialize Gyro functionality in your WebChucK instance.
+   * This adds a `Gyro` and `GyroMsg` class to the ChucK Virtual Machine (VM).
+   * Gyroerometer event (DeviceMotionEvent) listeners are added if `enableGyro` is true (default).
```
Comment (terryzfeng) [inline src/gyro/Gyro.ts:63]: nit: Gyrometer

File: src/gyro/Gyro.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * Enable Javascript event (DeviceMotionEvent) listeners for Gyro
+   * @example
+   * ```ts
+   * // If gyro is not yet enabled
+   * gyro.enableGyro();
+   * ```
+   */
+  enableGyro() {
+    // consider using "deviceorientationabsolute"
```
Comment (terryzfeng) [inline src/gyro/Gyro.ts:103]: Did we want to still consider this functionality down the line?

File: src/gyro/gyroCk.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    string deviceName; 
+
+    // GyroMsg Queue
+    GyroMsg _gyroMsgQueue[0];
+
+    function string name() {
+        return deviceName;
+    }
+
+    function int openGyro(int num) {
```
Comment (terryzfeng) [inline src/gyro/gyroCk.ts:49]: Just thinking out loud for the future here, no changes requested, but perhaps in the future, we should actually check if the gyro is working via js lol :)

File: src/gyro/gyroCk.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        if (num < 0) {
+            false => active;
+        } else {
+            "js DeviceOrientationEvent" => deviceName;
```
Comment (terryzfeng) [inline src/gyro/gyroCk.ts:53]: nit: Are we calling the Gyro device "js DeviceOrientationEvent" or is it like `js DeviceOrientationSensor`?

File: test/chuckTest.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 //=======================================================================
 // WebChucK Test Suite
 //=======================================================================
-import { Chuck, HID } from '../src/wc-bundle.js';
+import { Chuck, HID, Gyro, Accel} from '../src/wc-bundle.js';
```
Comment (terryzfeng) [inline test/chuckTest.js:4]: nit: import bracket spacing

File: src/accel/accelCk.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    function string name() {
+        return deviceName;
+    }
+
+    function int openAccel(int num) {
+        if (num < 0) {
+            false => active;
+        } else {
+            "js DeviceMotionEvent" => deviceName;
```
Comment (terryzfeng) [inline src/accel/accelCk.ts:53]: nit: Is the device name `js DeviceMotionEvent` or more like `js DeviceMotionSensor`?

File: src/accel/Accel.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    theChuck: Chuck,
+    enableAccel: boolean = true
+  ): Promise<Accel> {
+    const accel = new Accel(theChuck);
+    // Add Accel and AccelMsg classes to ChucK VM
+    await accel.theChuck.runCode(AccelMsg_ck);
+    await accel.theChuck.runCode(Accel_ck);
+
+    // Enable mouse and keyboard
+    accel.enableAccel();
```
Comment (terryzfeng) [inline src/accel/Accel.ts:79]: This line of code should only be run when `enableAccel` is true.

File: src/gyro/Gyro.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    theChuck: Chuck,
+    enableGyro: boolean = true
+  ): Promise<Gyro> {
+    const gyro = new Gyro(theChuck);
+    // Add Gyro and GyroMsg classes to ChucK VM
+    await gyro.theChuck.runCode(GyroMsg_ck);
+    await gyro.theChuck.runCode(Gyro_ck);
+
+    // Enable mouse and keyboard
+    gyro.enableGyro();
```
Comment (terryzfeng) [inline src/gyro/Gyro.ts:79]: This line of code should only be run when `enableGyro` is true.

File: src/accel/Accel.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+ * const accel = await Accel.init(theChuck); // Initialize Accel
+ * ```
+ * 
+ * The "devicemotion" event gives acceleration of the device on the three axes: x, y, and z. Acceleration is expressed in m/s².
+ * More on the "devicemotion" event can be found online {@link https://developer.mozilla.org/en-US/docs/Web/API/Window/devicemotion_event | here }.
+ * 
+ * iOS devices require that the web developer ask permission from the user to access sensors after a button push. For example: 
+ * 
+ * ```ts
+ * import { Chuck, Accel} from '../webchuck/src/wc-bundle.js';
```
Comment (terryzfeng) [inline src/accel/Accel.ts:25]: nit: import bracket spacing and can just make it `import { Chuck, Accel } from "webchuck"` like above

**Review by terryzfeng — APPROVED** [review]
LGTM! If you could fix the `enableGyro/Accel` stuff that'd be great! Others are just some small documentation/nitpicks ;)

Thanks for making these two classes happen Mike!

File: src/gyro/Gyro.ts (hunk 1 above)
Comment (mulshine) [reply src/gyro/Gyro.ts:6]: ah yes, gyroscope*

File: src/gyro/Gyro.ts (hunk 4 above)
Comment (mulshine) [reply src/gyro/Gyro.ts:103]: I don't think so. It isn't supported on some browsers, and seems deviceorientation is the more supported variety.

File: src/gyro/gyroCk.ts (hunk 5 above)
Comment (mulshine) [reply src/gyro/gyroCk.ts:49]: Yes, that would be smart. Still some things to sort or keep an eye on there.

File: src/accel/accelCk.ts (hunk 8 above)
Comment (mulshine) [reply src/accel/accelCk.ts:53]: I don't think there is a specific device name... it's just representing access to device orientation and acceleration. I made up this shorthand reference... we could just call it "accelerometer" or "gyroscope" as that is essentially what we are accessing. What do you think?

**Comment by terryzfeng** [comment]
LGTM!

**Pull request merged**